File: Args.java
/*

The MIT License (MIT)

Copyright (c) <2012-2016> eli.shagam@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy of this software
and associated documentation files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished 
to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE
AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

 */

package mpprunner;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import java.util.TreeMap;

/**
 * Jewel Store Copyright (C) 2011-2014  TechoPhil Ltd
 *
 * @author Eli Shagam
 *
 *  retrieve invocation parameters
 * 

 *   usage
 * 
 *      See utest()  below
 *
 * 
 */
public class Args {
    
    static final int ARGS_MAX = 20;
    static ArrayList <String> s_searchedParams = new ArrayList<> ();
    static Map <String, String> s_descriptionMap = new TreeMap <> ();
    static ArrayList <ArgType> s_argType = new ArrayList<> ();
    static String [] s_args;
    public static String s_err = null;
    static String [] [] s_enumHelp;
    
    enum ArgType {
        bool,
        string,
        integer,
        enumer,
    }
    

    
    // search one formalParam in actual args
    public static String getString (String name, String [] args, String description) {
        if (s_args == null)
            s_args = args;
        String param = extractParam (name);        
        checkDuplicatesAndAddToList(param);
        s_descriptionMap.put (name, description);
        s_argType.add (ArgType.string);
        for (int n = 0; n < args.length; n++) {
            // remove leading dashes
            String arg = args[n];
            if (arg.startsWith("--"))
                arg = arg.substring(2);            
            if (arg.contains("=")) {
                String [] array = arg.split("=");
                if (array.length != 2)
                    continue;
                if (! matchPrefix (name, array[0]))
                    continue;
                return array[1];
            }
            if (matchPrefix (name, args[n]))
                s_err = "miss_value " + args[n];
                    
        }
        return null;
    }
      
    public static int getInteger (String name, String [] args, String description) {
        String strValue = getString (name, args, description);
        s_argType.remove(s_argType.size() - 1);
        s_argType.add (ArgType.integer);
        s_descriptionMap.put (name, description);
        if (strValue == null)
            return Integer.MAX_VALUE;
        
        int val;
        try {
            val = Integer.parseInt(strValue, 10);
            return val;
        }
        catch (NumberFormatException e) {
            s_err = name + " invalid int";
            System.err.print(name + " invalid int" );
            return Integer.MAX_VALUE;
        }
    }

    public static boolean getBool (String name, String [] args, String description) {
        if (s_args == null)
            s_args = args;

        checkDuplicatesAndAddToList (name);
        s_descriptionMap.put (name, description);
        s_argType.add (ArgType.bool);
        for (int n = 0; n < args.length; n++) {
            // remove leading dashes
            String arg = args[n];
            if (arg.startsWith("--"))
                arg = arg.substring(2);
            
            if (matchPrefix(arg, name)) {
                if (arg.contains("="))
                    s_err = "misMatchType " + arg;
                return true;       
            }
        }
        return false;
    }
    
    public static String getEnum (String name, String [] args, String[] enumerator, String [] enumHelp, String grossDescription) {
        if (s_args == null)
            s_args = args;
        //checkDuplicatesAndAddToList (name);

        grossDescription += " enum alternatives: ";
        for (int n = 0; n < enumerator.length; n++) {
            grossDescription += enumerator[n];
            if (n < enumerator.length - 1)
                grossDescription += "|";
        }
        
        String selection = getString (name, args, grossDescription);
        s_argType.remove(s_argType.size() - 1);
        s_argType.add (ArgType.enumer);          

        
        if (s_enumHelp == null) {
            s_enumHelp = new String [ARGS_MAX][];
        }
        assert s_searchedParams.size() < ARGS_MAX;
        s_enumHelp [s_searchedParams.size() - 1] = enumHelp;
        
        if (selection == null)
            return null;
                 
        assert enumerator.length == enumHelp.length;
         String result = null;
         int count = 0;
         String alternatines = "";
         for (int n = 0; n < enumerator.length; n++) {
             if (enumerator[n].startsWith(selection)) {
                 count++;
                 result = enumerator[n];
                 alternatines += enumerator[n] + ",";
             }
         }
         switch (count) {
            case 1:
                return result;
            case 0:
                 return null;
            default:
                System.err.print("\nmultiple match: " + alternatines + "\n");
            return null;
         }
     }

    
    static void checkDuplicatesAndAddToList (String name) {
        String param = extractParam (name);
        if (param == null) {
            s_err = " invalid " + name;
            return;
        }
        for (int n = 0; n < s_searchedParams.size(); n++) {
            if (matchPrefix(param, s_searchedParams.get(n))) {
                s_err = " duplicate=" + param;
                return;
            }
        }
        s_searchedParams.add(name);         
    }

    static String extractParam (String nameValue) {
        if (! nameValue.contains("="))
            return nameValue;
        String [] array = nameValue.split("=");
        if (array.length == 2)
            return array[0];
        else
            return null;
    }
    
    static boolean matchPrefix (String a, String b) {
        int len = min (a.length(), b.length());
        a = a.substring(0, len);
        b = b.substring(0, len);
        return a.matches(b);
    }
   
    public static void showAndVerify (boolean help) {
        assert Args.s_err == null : s_err; 
        String txt = "";

        if (help) {
            txt += optionalParams();
        }
        
        //txt += actualArgs();

        System.err.print (txt); // print optional parameters        

        String err =  Args.verifyArgs();
        assert err == null : optionalParams() + err;
        verifyUnique ();        
        assert Args.s_err == null : s_err;
    }
    
    public static String actualArgs () {
        if (s_args.length == 0)
            return "";
        String txt = "\nactual args:  ["; 
        for (int n = 0; n < s_args.length; n++) {
            txt += s_args[n] + ", ";
        }
        txt += "]\n\n";
        return txt;
    }

    public static String optionalParams () {
        final int TXT_LEN = 30;        
        String txt = "";
        txt += "optional args:   ";
        for (int n = 0; n < s_searchedParams.size(); n++) {
            String txt1 = "\n";
            String arg = s_searchedParams.get(n);
            txt1 += arg;
            if (s_argType.get(n) == ArgType.bool)
                txt1 += "(" + s_argType.get(n) + ")  ";
            else
                txt1 += "=<" + s_argType.get(n) + ">  ";

            if (txt1.length() < TXT_LEN)
                txt1 += "                                            ".substring(0, TXT_LEN - txt1.length());
            int len = txt1.length();

            String descr = s_descriptionMap.get(arg);
            assert descr != null;
            txt1 += "        " + descr;
            
            if (s_argType.get(n) == ArgType.enumer) {
                for (String help :s_enumHelp[n]) {
                    txt1 += "\n       " + help;
                }                
            }
            txt += txt1;
        }
        return txt;
    }
        
    // verify all args appear in search list
    public static String verifyArgs () {
        for (int nArg = 0; nArg < s_args.length; nArg++) {
            boolean found = false;
            String arg = extractParam(s_args[nArg]);
            if (arg.startsWith("--"))
                arg = arg.substring(2);
            for (int param = 0; param < s_searchedParams.size(); param ++) {
                String parameter = s_searchedParams.get(param);
                if (matchPrefix (parameter, arg)) {
                    found = true;
                    break;
                }
            }
            if (! found)
                return "  not found param=" + arg;
        }
        return null;
    }
    
    public static void exit (String str, int err) {
        System.err.format("\n" + str + "\n");
        System.exit (err);        
    }
   
    // check if actual arg matches multiple params
    public static void verifyUnique () {
        for (int nArg = 0; nArg < s_args.length; nArg++) {
            String arg = s_args[nArg];
            if (arg.startsWith("--"))
                arg = arg.substring(2);
            String argName;
            if (arg.contains("=")) {
                String [] array = arg.split("=");
                argName = array[0];
            }
            else
                argName = arg;
            
            Collection <String> matches = new ArrayList<> ();
            for (int n = 0; n < s_searchedParams.size(); n++) {
                String param = s_searchedParams.get(n);

                if (matchPrefix (param, argName))
                    matches.add (param);
                assert matches.size() <= 1 : "duplicate matches of arg \"" + argName + "\" params: " + matches;
            }
        }
        
    }

    public static int min (int a, int b) {
        if (a <= b)
            return a;
        else
            return b;
    }
    
    // utest can be used as a demo
    public static void utest () {
        String [] args = {"th=7", "test=copy", "file=/var/js/blabla", "verbose"};

        // Args.getBool 
        boolean verbose = Args.getBool ("verbose", args, "");
        assert verbose;
        
        // Args.getString
        String file = Args.getString ("file", args, "file name");
        assert file != null;
        
        // Args.getInteger
        int threadCount = -1;
        //Args.search_int_value("thredd", args, "number of concurrent threads");        
        int threads = Args.getInteger("threds", args, "number of concurrent threads");
        if (threads != Integer.MAX_VALUE)
            threadCount = threads;
        else
            threadCount = Runtime.getRuntime().availableProcessors();
        assert threadCount == 7;

        // Args.getEnum 
        String [] args1 = {"test=tre"};        
        String [] tests = {"tree","hash","queue","alloc","prime","queue","copy","count","noLock"};
        String test = Args.getEnum ("test", args, tests, tests, "test selection ");
        
        // verify and print
        Args.showAndVerify (true);
        System.exit(0);
    }
}

File: args.c


/*

The MIT License (MIT)

Copyright (c) <2012-2016> eli.shagam@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy of this software
and associated documentation files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished 
to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE
AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

 */



#include <iostream>
#include "args.h"
#include <string>
#include <map>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

//#include <string.h>

using namespace std;

//static char * s_argv[];

const int STR_MAX = 100;
const int ARGS_MAX = 20;

const char * paramNameArray[ARGS_MAX];
PARAM_TYP paramTypeArray[ARGS_MAX];
const char * param_type_name[ARGS_MAX];
const char * paramDescriptionArray[ARGS_MAX];
const char * * s_enumHelp [ARGS_MAX];

int paramCount = 0;
int s_enumHelpCount;

int isPrefix(char const * prefix, char const * str)
{
    int p_len = strlen (prefix);
    int q_len = strlen (str);
    
    if (p_len > q_len)
        return 0;
                
    while (*prefix && *str) {
        if (*prefix++ != *str++)
            return 0;
    }
    return 1;
}

int isContain (char const *p, char c) {

    while (*p) {
        if (*p++ == c)
            return 1;
    }
    return 0;    
}

int isNumeric (const char* str) {
    
    for (int n = 0; n < sizeof(str) - 1; n++) {
        if (str[n] == 0)
            return 1;
        if (! isdigit (str[n])) {
            fprintf (stderr, "\nbad numeric value=%s %s", str, "\n");            
            exit (1);
        }
    }
    return 1;
}

int getBool (char const * name, int argc, const char * const argv[], char const * description) {
    paramNameArray[paramCount] = name;
    paramTypeArray[paramCount] = BOOL;
    param_type_name[paramCount] = "bool";
    paramDescriptionArray[paramCount] = description;    
    paramCount ++;
    for (int n = 1; n < argc; n++) {
        if (isContain(argv[n],'='))
            continue;    
        if (isPrefix (argv[n], name)) {           
            return 1;
        }
    }
    return 0;
}

int split (char const * name_value, char * param_name, char * value) {
    int len = strlen (name_value);
    for (int n = 0; n < len; n++) {
        if (*(name_value + n) == '=') {
            if (len - n - 1 > 0) {
                memcpy (param_name, name_value, n);
                memcpy (value, name_value + n + 1, len - n - 1);
            }
            else {
                param_name[0] = 0;
                value[0] = 0;
            }
                
            return 1;
        }
    }
    return 0;    
    
}

const char * getString (char const * param_name, int argc, const char * const argv[], char const * description) {
    paramNameArray[paramCount] = param_name;
    paramTypeArray[paramCount] = STRING;
    param_type_name[paramCount] = "string";
    paramDescriptionArray[paramCount] = description; 
    paramCount ++;   
    for (int n = 0; n < argc; n++) {
        const char * e = strchr(argv[n], '=');
        if (e == NULL)
            continue;
        e++; // skip =
        if (e - argv[n] >= strlen (argv[n]))
            return NULL;
        {
            char name [STR_MAX] = {0};
            char value [STR_MAX] = {0};    

            int res = split (argv[n], name, value);
            if (res = 0)
                continue;

            int nameLen = strlen(name);
            int valueLen = strlen (value);

            if (nameLen == 0 || valueLen == 0)
                continue;

            if ( ! isPrefix(name, param_name))
                continue;
            //fprintf (stderr, "\nsearch_value=%s\n", value);            
            return e;
        }
    }
    return NULL;
}

int getInteger (char const * param_name, int argc, const char * const argv[], char const * description) {
    
    const char * res = getString (param_name, argc, argv, description);
    paramCount --;
    paramNameArray[paramCount] = param_name;
    paramTypeArray[paramCount] = INT;
    param_type_name[paramCount] = "int";    
    paramDescriptionArray[paramCount] = description;      
    paramCount ++;
    
    if (res == NULL)
        return -1;

    if (! isNumeric(res)) {
        //fprintf (stderr, "\nbad numeric value \n");
        exit (1);
    }
    int value = atoi(res);    

    return value;
}

const char* argsGetEnum (const char* name, int argc, const char * const args[], int enum_count,
        const char * enumerator[], const char * enumHelp[], char const * description) {
    s_enumHelp[paramCount] = enumHelp;
    s_enumHelpCount = enum_count;
    int ptr = 0;
    char descr[STR_MAX] = {0};
    ptr += sprintf (descr, "%senum alternatives: ", description);
    for (int n = 0; n < enum_count; n++) {
        ptr += sprintf (descr + ptr, "%s", enumerator[n]);
        if (n < enum_count - 1)
            ptr += sprintf (descr + ptr, "|");
    }

    const char * selection = getString (name, argc, args, descr);
    paramCount --;
    paramNameArray[paramCount] = name;
    paramTypeArray[paramCount] = ENUM;
    param_type_name[paramCount] = "enum";    
    paramDescriptionArray[paramCount] = description;   
    paramCount ++;    
    if (selection == NULL)
         return NULL;
  
     
     const char * result1;
     int count = 0;
     char alternatines[STR_MAX] = {0};
     ptr = 0;
     for (int n = 0; n < enum_count; n++) {
         if (isPrefix(selection, enumerator[n])) {
             count++;
             result1 = enumerator[n];
             ptr += sprintf (alternatines + ptr, "%s,", enumerator[n]);
         }
     }
     switch (count) {
        case 1:
            return result1;
        case 0:
             return NULL;
        default:
            fprintf (stderr, "\nmultiple match: %s\n", alternatines);
        return NULL;
     }
 }


void args_report () {
    char report [1000] = {0};
    int ptr = 0;
    std::cerr << "\nparams:  ";
    for (int n = 0; n < paramCount; n++) {
    //for(auto elem : param_map) {
        const int GAP = 20;
        int len;
     
        len = strlen (paramNameArray[n]);
        switch (paramTypeArray[n]) {
            case  BOOL:
            printf ("\n%s(%s)  ", paramNameArray[n], "bool");
            len += 2;
            break;
            
            case STRING:
            printf ("\n%s=<%s>", paramNameArray[n], "string");
            break;

            case INT:
            printf ("\n%s=<%s>", paramNameArray[n], "int");
            break;
            
            case ENUM:
            printf ("\n%s=<%s>", paramNameArray[n], "enum");
            break;
        }        
        if (len < GAP) {
            if (paramTypeArray[n] == STRING)
                len += 3;
            for (int n = 0; n < GAP - len; n++)
                printf (" ");
        }
        
        printf ("      %s", paramDescriptionArray [n]);
        
        if (paramTypeArray[n] == ENUM) {

            for (int i = 0; i < s_enumHelpCount; i++) {
                printf ("\n        %s", s_enumHelp[n][i]);                
            }
        }
        
    }
    std::cerr << "\n";
}

int formatInt (char* str, long n) {
        const int BILION = 1000000000L; 
        const int MILION = 1000000L; 
        const int KILO = 1000L; 
        int fill = 0;
        char* str_save = str;

        if (n > BILION) {
            str += sprintf(str, "%ld,", n / BILION);
            n -= n / BILION * BILION;
            fill++;
        }
        if (n > MILION) {
            if (fill)
                str += sprintf(str, "%03ld,", n / MILION);
            else
                str += sprintf(str, "%ld,", n / MILION);
            n -= n / MILION * MILION;
            fill++;
        }        
        if (n > KILO) {

            if (fill)
                str += sprintf(str, "%03ld,", n / KILO);
            else
                str += sprintf(str, "%ld,", n / KILO);
            n -= n / KILO * KILO;
            fill++;          
        }
        if (fill)
            str += sprintf(str, "%03ld", n);
        else
            str += sprintf(str, "%ld", n);
        return str - str_save;
    }
    
void formatIntUtest (void) {
    char str1 [1000] = {0};
    char* str = str1;
    str += sprintf (str, "\n");
    str += formatInt(str, 3333034567089L);
    str += sprintf (str, "\n");
    str += formatInt(str, 4567089);
    str += sprintf (str, "\n");
    str += formatInt(str, 7089);
    str += sprintf (str, "\n");
    str += formatInt(str, 89);
    str += sprintf (str, "\n");
    printf ("%s", str1);
    
}

// verify valid args 
const char * verify (int argc, const char * const argv[]) {
    for (int nArg = 1; nArg < argc; nArg++) {
        const char * nameValue = argv[nArg];
        if (nameValue == NULL)
            return NULL;       
        char argName [STR_MAX] = {0};
        char value[STR_MAX] = {0};
        const char * e = strchr(nameValue, '=');
        if (e == NULL) {
            memcpy (argName, nameValue, strlen (nameValue));
        }
        else
            int res = split (nameValue, argName, value);
        int found = 0;
        
        for (int param = 0; param < paramCount ; param ++) {
            const char * parameter = paramNameArray[param];

            if (isPrefix (argName, parameter)) {
                found = 1;
                break;
            }
        }
        if (! found)
            return /* "  invalid param=" + */ argv[nArg];
    }
    return NULL;
  
}

void utest () {
       
    static const char *const argv[] = {"testNam=allo","delay=4", "help"};
    static const int argc = 3;

    // get string param
    const char * status = getString ("testName", argc, argv, "description: type of test");
    if (status == NULL) { // means param testName is missing 
        fprintf (stderr, "\n  stat=%s\n", status);
    }

    // get int param
    int delay = getInteger ("delay", argc, argv,"description: duration of test");
    if (delay == -1) {  //-1  means delay arg missing
        fprintf (stderr, "\nvalue_int=%d\n", delay);
    }
    
    const char * tests[] = {"tree","hash","queue","alloc","prime","copy","count","noLock"};
    const char * testHelp[] = {"tree retrieve", "hash retrieve", "queue get/enque", "alloc new / delete (With variable arraySize)",
            "prime compute intensive", "copy mem (Bandwidth)", "common counter increment by many threads", "tree retrieve without lock"};    
    
    const char * test = argsGetEnum ("testName", argc, argv, 8, tests, testHelp, "test selection ");
    
    // get boolean param 0/1
    int help = getBool ("help", argc, argv,"description: print optional params");
    if (help)
        args_report(); // if help exist print all allowed arguments;  arg_report after all args collected
    
    // verify no bogus arguments
    const char * err = verify (argc, argv);
    if (err != NULL) {
        fprintf (stderr, "invalid param=%s  \n", err);
        exit (3);
    }
}

// test for sub routines
int utest_sub () {
   int res = isContain ("aaa=bbbb", '=');
//    if (res)
//        fprintf (stderr, "contain\n");
    
    res = isPrefix ("aaa", "aaattt\n");
    if (res)
        fprintf (stderr, "prefix\n");
    
    
    
    //const int SIZ = 20;
    char name_value [] = "param=value";
    char name [STR_MAX];
    char value [STR_MAX];    
    
    res = split (name_value, name, value);
    int nameLen = strlen(name);
    int valueLen = strlen (value);
    
    if (res && valueLen > 0) {
        fprintf (stderr, "\nSplit   in:%s  name=%s %d   value=%s %d\n", name_value, name, nameLen, value, valueLen);
    }
    
    static const char *const days[] = { "mon", "tue", "wed", "thur",
                                       "fri", "sat", "sun" };
    
}


int getCPUCount() {
 cpu_set_t cs;
 CPU_ZERO(&cs);
 sched_getaffinity(0, sizeof(cs), &cs);

 int count = 0;
 for (int i = 0; i < 20; i++)
 {
  if (CPU_ISSET(i, &cs))
   count++;
 }
 return count;
}